lang/LangPrimSource/PyrArrayPrimitives.cpp

```diff
@@ -2474,6 +2477,95 @@ int prArrayUnlace(struct VMGlobals* g, int numArgsPushed) {
     return errNone;
 }
 
+int arrayLevenshteinDistance(struct VMGlobals* g, PyrSlot* result, PyrObject* thisArray, PyrObject* thatArray) {
+    int distance = sc_max(thisArray->size, thatArray->size); // assume they're completely different
+
+    if (thisArray->size == 0) {
+        distance = thatArray->size;
+        SetInt(result, distance);
+        return errNone;
+    }
+
+    if (thatArray->size == 0) {
+        distance = thisArray->size;
+        SetInt(result, distance);
+        return errNone;
+    }
+
+    // if they're both the same format we can compare them..
+    if (thisArray->obj_format == thatArray->obj_format) {
+        switch (thisArray->obj_format) {
+        // if they're raw arrays, compare as such
+        case obj_char:
+            distance = levenshteinDistance<char>()(reinterpret_cast<char*>(thisArray->slots), thisArray->size,
+                                                   reinterpret_cast<char*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_double:
+            distance = levenshteinDistance<double>()(reinterpret_cast<double*>(thisArray->slots), thisArray->size,
+                                                     reinterpret_cast<double*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_float:
+            distance = levenshteinDistance<float>()(reinterpret_cast<float*>(thisArray->slots), thisArray->size,
+                                                    reinterpret_cast<float*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_int16:
+            distance = levenshteinDistance<int16>()(reinterpret_cast<int16*>(thisArray->slots), thisArray->size,
+                                                    reinterpret_cast<int16*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_int32:
+            distance = levenshteinDistance<int32>()(reinterpret_cast<int32*>(thisArray->slots), thisArray->size,
+                                                    reinterpret_cast<int32*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_int8:
+            distance = levenshteinDistance<int8>()(reinterpret_cast<int8*>(thisArray->slots), thisArray->size,
+                                                   reinterpret_cast<int8*>(thatArray->slots), thatArray->size);
+            break;
+        case obj_slot: {
+            // if it's slotted, compare identities
+            // this will handle polymorphic arrays in a predictable way
+            auto calcDistance = levenshteinDistance<PyrSlot>();
+
+            distance = calcDistance(thisArray->slots, thisArray->size, thatArray->slots, thatArray->size,
+                                    [](const PyrSlot& a, const PyrSlot& b) {
+                                        return SlotEq(const_cast<PyrSlot*>(&a), const_cast<PyrSlot*>(&b));
```

Comment:
yes - this was only to avoid changing SlotEq, which for some reason isn't const. 

I didn't want to modify such a critical component to suit my needs, but more than happy to do it? It should've been const the whole time, really.

Am I marking the function itself as const, or just the arguments?

Reply:
> I didn't want to modify such a critical component to suit my needs, but more than happy to do it? It should've been const the whole time, really.

yes, exactly. i can assure you that marking the parameters const will have absolutely no effect on the correctness of any code that uses it.

> Am I marking the function itself as const, or just the arguments?

`const` applied to functions is only possible when the function is a member of a class, where it indicates that none of the members of the class (except for those designated `mutable`) may be modified inside the function. since `SlotEq` is a free function (i.e., not a member of a class), you can and should only mark the arguments `const PyrSlot*` instead of `PyrSlot*`.